File: Python/GraphTheory/Dijkstra.py
from heapq import heappush, heappop
INF = 998244353
 
def dijkstra(s,n,adj):
    dist = [INF] * n
    dist[s] = 0
    h = [(0,s)]
 
    while h:
        d,v = heappop(h)
        if d > dist[v]:
            continue
 
        for to,cost in adj[v]:
            if dist[v] + cost < dist[to]:
                dist[to] = dist[v] + cost
                heappush(h,(dist[to],to))
    return dist
 
'''verify: https://atcoder.jp/contests/joi2008yo/tasks/joi2008yo_f

n,k = map(int,input().split())
adj = [[] * n for _ in range(n)]
 
for i in range(k):
    l = list(map(int,input().split()))
    if l[0] == 0:
        res = dijkstra(l[1] - 1,n,adj)
        if res[l[2] - 1] == INF:
            print(-1)
        else:
            print(res[l[2] - 1])
    else:
        adj[l[1] - 1].append((l[2] - 1,l[3]))
        adj[l[2] - 1].append((l[1] - 1,l[3]))
'''

File: C++/GraphTheory/Dijkstra.cpp
void dijkstra(int n) {
	priority_queue<pair<ll, ll>, vector<pair<ll, ll>>, greater<pair<ll, ll>>>que;
	fill(dij.begin(), dij.end(), INF);
	dij[n] = 0;
	que.push(pair<long long, long long>(0, n));

	while (!que.empty()) {
		pair<long long, long long>p = que.top();
		que.pop();
		long long mon = p.second;
		if (dij[mon] < p.first) continue;
		for (int i = 0; i < G[mon].size(); ++i) {
			edge e = G[mon][i];
			if (dij[e.to] > dij[mon] + e.cost) {
				dij[e.to] = dij[mon] + e.cost;
				que.push(pair<long long, long long>(dij[e.to], e.to));
			}
		}
	}
}

/*verify: https://atcoder.jp/contests/joi2008yo/tasks/joi2008yo_f
int main() {
	int n, k;
	cin >> n >> k;
	dij.resize(n);
	G.resize(n);
	for (int i = 0; i < k; ++i) {
		int d;
		cin >> d;
		if (d) {
			int a, b, c;
			cin >> a >> b >> c;
			--a, --b;
			addedge(a, b, c);
		}
		else {
			int a, b;
			cin >> a >> b;
			--a, --b;
			dijkstra(a);
			cout << (dij[b] == INF ? -1 : dij[b]) << endl;
		}
	}
}
*/